lahja/asyncio/endpoint.py

```diff
@@ -759,6 +759,8 @@ async def stream(
                     yield await queue.get()
                 except GeneratorExit:
                     break
+                else:
+                    await asyncio.sleep(0)
```

Comment:
The `await queue.get()` above should always yield back to the event loop

Reply:
So I decided it was time to finally learn some more about asyncio. So now I'll be a little pedantic, because I've misunderstood this idea that "await yields to the event loop." It can, but there are plenty of times it won't. So a slightly more precise version in this case is:

> The `await queue.get()` above ~should always~ *will* yield back to the event loop, *but only when the queue is empty*


This modified example from [this stackoverflow post](https://stackoverflow.com/a/59780868/8412986) helped me understand it a bit more:
```py
import asyncio
from itertools import count

counter = count()


def task_print(message):
    print(message)


async def async_print(message):
    task_print(message)


async def next_val():
    return next(counter)


async def arange(size):
    for _ in range(size):
        yield await next_val()


async def original_task():
    asyncio.create_task(async_print("This doesn't run until the very end because the generated None following the creation of this task is consumed by the `for` loop"))
    async for y in arange(3):
        await async_print(f"'ordinary' `yield` points (those which are consumed by the `current_task` itself) behave as ordinary without relinquishing control at the async/task-level; `y={y}`")
    task_print("Done with original task")

asyncio.get_event_loop().run_until_complete(original_task())
```
Prints:

> 'ordinary' `yield` points (those which are consumed by the `current_task` itself) behave as ordinary without relinquishing control at the async/task-level; `y=0`
> 'ordinary' `yield` points (those which are consumed by the `current_task` itself) behave as ordinary without relinquishing control at the async/task-level; `y=1`
> 'ordinary' `yield` points (those which are consumed by the `current_task` itself) behave as ordinary without relinquishing control at the async/task-level; `y=2`
> Done with original task
> This doesn't run until the very end because the generated None following the creation of this task is consumed by the `for` loop

The number of awaits doesn't matter, and it won't release at the beginning or middle or end of an await, even multiple layers down. Only when we await something that then `yield`s to the event loop.

Reply:
Less thorough than @carver but I did a test before and was surprised to find it didn't yield (when there were items in the queue).

Reply:
Yeah, it makes sense that `await queue.get()` will only relinquish control to the loop (trying to avoid the term yield as we're talking about generators) when the queue is empty, so I guess the question we need to answer now is whether or not we want `stream()` to relinquish control to the loop after every item yielded? My gut feeling is that it'd make more sense to not worry about that here (but document its behavior clearly) and let consumers decide how often they want to relinquish control. The main reason being that I suspect in most cases the number of events streamed is relatively small so that we can stream them all in one go. And if a consumer is doing expensive operations for each of those events, it should be doing so in a cooperative fashion, which would relinquish control back to the event loop. It's also important to notice that when the queue is empty it will relinquish control one extra time, unnecessarily.

Reply:
I couldn't find any async iterators/generators in the stdlib so that we could at least be consistent with, but the websocket library has one which will yield all queued messages before relinquishing control to the loop: https://github.com/aaugustin/websockets/blob/master/src/websockets/protocol.py#L439-L523

Reply:
I just checked and `trio.ReceiveChannel` used as an async iterator yields to other tasks on each item.  I'd advocate following their lead (which is what this PR does).

Reply:
Fair enough

Reply:
So factors at play:
- added cost of yielding to the event loop in between each event (I'm not sure what the order of magnitude cost is here)
- risk of holding the event loop too long
- when several events come through simultaneously, the later messages will have higher latency because of yielding control

It's not an obvious answer here, especially because of the unknown cost of yielding control. My instinct says that we should yield at every element here. It's really difficult to detect and track down where the event loop is holding control for too long (at least in my install, for some reason). It can have pretty dramatic effects on performance.

One small suggestion: we can avoid yielding to the event loop twice by checking if the queue is empty before the `.get()`. If it was, then we can skip the `sleep(0)`.

Talking to myself: it would be kind of neat to be able to call `loop.time_since_yield()` to know how long the current coro has been holding control of the loop.

Reply:
I'll update to do the more intelligent conditional `sleep(0)` based on whether we've already yielded due to the queue being empty.